src/ui_fsmenu/mousewheel_options.cc

```diff
@@ -0,0 +1,438 @@
+/*
+ * Copyright (C) 2021 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#include "ui_fsmenu/mousewheel_options.h"
+
+#include <list>
+
+#include <SDL_keycode.h>
+#include <boost/format.hpp>
+
+#include "base/i18n.h"
+#include "graphic/font_handler.h"
+#include "graphic/text_layout.h"
+#include "ui_basic/box.h"
+#include "ui_basic/button.h"
+#include "ui_basic/dropdown.h"
+#include "ui_basic/messagebox.h"
+#include "ui_basic/textarea.h"
+#include "ui_fsmenu/menu.h"
+#include "wlapplication_mousewheel_options.h"
+#include "wlapplication_options.h"
+
+namespace FsMenu {
+
+constexpr int kButtonSize = 24;
+constexpr int kDividerSpace = 8;
+constexpr int kModDirDropdownMaxWidth = 200;
+constexpr int kApplyButtonWidth = 400;
+constexpr int kMousewheelBoxMaxWidth = 700;
+
+// Scroll Directions
+enum SD : uint8_t {
+	kDisabled = 0,
+	kNeither = kDisabled,
+	kVertical = 1,
+	kHorizontal = 2,
+	kAny = kVertical | kHorizontal,
+	kBoth = kAny
+};
+
+static const std::string sd_names[] = {
+   /** TRANSLATORS: Placeholder for scroll direction when function should not use scroll wheel */
+   gettext_noop("Disabled"),
+   /** TRANSLATORS: e.g. Ctrl+Vertical scroll */
+   gettext_noop("Vertical scroll"),
+   /** TRANSLATORS: e.g. Ctrl+Horizontal scroll */
+   gettext_noop("Horizontal scroll"),
+   /** TRANSLATORS: Used when a function works with either vertical or horizontal scrolling.
+       It may be used in combination with modifier keys, e.g. "Ctrl+Any scroll" */
+   gettext_noop("Any scroll")};
+
+#define READ_MOD(option) normalize_keymod(get_mousewheel_keymod(MousewheelOptionID::option##Mod))
+
+#define DIR_COMBINE(x, y)                                                                          \
+	((x ? SD::kHorizontal : SD::kDisabled) | (y ? SD::kVertical : SD::kDisabled))
```

Comment:
This macro can be replaced with an inline function:
```suggestion
inline uint8_t dir_combine(const bool x, const bool y) {
	return ((x ? SD::kHorizontal : SD::kDisabled) | (y ? SD::kVertical : SD::kDisabled));
}
```